### ISSUE TITLE
Cannot use packages "System.Runtime.Loader", "System.Reflection.Emit" and "System.Reflection.Emit.Lightweight" when targeting netstandard2.0

### ISSUE THREAD

**Issue by daxian-dbw** [dotnet/corefx#17746]
I’m moving [PowerShell Core](https://github.com/powershell/powershell) to NetStandard2.0, and find that I cannot depend on the following packages when building an assembly targeting netstandard2.0:
- [“System.Runtime.Loader”](https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Runtime.Loader)
- [“System.Reflection.Emit”](https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Reflection.Emit)
- [“System.Reflection.Emit.Lightweight”](https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Reflection.Emit.Lightweight)

I’m targeting a very recent version of NetStandard.Library package (2.0.0-beta-25123-01), where the core/primitive types like `Systme.Object` have been moved to netstandard.dll. However, the newest version of the above 3 packages on [dotnet.myget](https://dotnet.myget.org/gallery/dotnet-core) is 4.4.0-beta-24913-02, which is 3 months old, and thus they are still on top of `System.Runtime.dll`. When listing them in the package references, `dotnet restore` will bring the old version of System.Runtime.dll as a dependency, and then `dotnet build` will throw following ambiguous type identity errors:

```output
AssemblyInfo.cs(3,12): error CS0433: The type 'InternalsVisibleToAttribute' exists in both 'System.Runtime, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' and 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' [D:\PowerShell\src\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.csproj]
AssemblyInfo.cs(4,12): error CS0433: The type 'AssemblyFileVersionAttribute' exists in both 'System.Runtime, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' and 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' [D:\PowerShell\src\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.csproj]
AssemblyInfo.cs(5,12): error CS0433: The type 'AssemblyVersionAttribute' exists in both 'System.Runtime, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' and 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' [D:\PowerShell\src\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.csproj]
```

I know the types/APIs exposed from those 3 packages are not in NetStandard2.0, but the packages should be on top of NetStandard and I should be able to use them when targeting NetStandard2.0, right?
What should I do to continue depending on those 3 packages when targeting NetStandard2.0?

## Environment data
```powershell
PS:515> dotnet --info
.NET Command Line Tools (2.0.0-preview1-005683)

Product Information:
 Version:            2.0.0-preview1-005683
 Commit SHA-1 hash:  6866bc2f4a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   X:\dotnet\dotnet-dev\sdk\2.0.0-preview1-005683\
```

**Comment by danmosemsft** [comment]
@weshaggard ?

**Comment by weshaggard** [comment]
This is by design. They only support .NET Core, so if you want to consume them you need to do it from a netcoreapp library/application.

**Comment by daxian-dbw** [comment]
Thanks @weshaggard! If my assembly library is targeting `netcoreapp2.0`, can it be consumed by a library project that targets netstandard2.0?

**Comment by karelz** [comment]
I am not sure if it is technically achievable (if nothing prevents you), but the key point is that it doesn't make much sense:
If your library targets netstandard2.0, then its purpose is to work across multiple platforms (.NET Core, Desktop, Xamarin, UWP). If it depends also on platform-specific lib (netcoreapp20), it will loose the multiple-platformness. So you might as well make it target netcoreapp20 as well.

cc @terrajobst - do we have docs explaining these things?

**Comment by daxian-dbw** [comment]
@karelz Thanks for the clarification!
I ran into another issue when using the package [`Microsoft.CodeAnalysis.CSharp/2.0.0`](https://www.nuget.org/packages/Microsoft.CodeAnalysis.CSharp/) (_2.0.0 is the latest version I can find_) in my project targeting `netcoreapp2.0`. `Microsoft.CodeAnalysis.CSharp/2.0.0` depends on `Microsoft.CodeAnalysis.Common/2.0.0` which pulls in `System.ValueTuple/4.3.0`. However, the types in `System.ValueTuple.dll` are already exposed from `System.Runtime.dll` when targeting `netcoreapp2.0`, and thus using any of those types results in ambiguous type identity error:
```
Program.cs(9,45): error CS0433: The type 'TupleExtensions' exists in both 'System.ValueTuple, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' and 'System.Runtime, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' [D:\apps\netcore2\test\test.csproj]
```
Here is a simple repro:
```
using System;

namespace test
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine(typeof(System.TupleExtensions).FullName);
        }
    }
}

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <RuntimeIdentifier>win10-x64</RuntimeIdentifier>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="2.0.0" />
  </ItemGroup>

</Project>
```

It looks `netcoreapp2.0` doesn't come with `Microsoft.CodeAnalysis.CSharp.dll` and `Microsoft.CodeAnalysis.dll`. How should I consume the `Microsoft.CodeAnalysis.CSharp` package?

**Comment by karelz** [comment]
@daxian-dbw that's an interesting problem on its own, can you please file separate issue? (and delete/update this reply?) It doesn't help to mingle them into one issue, esp. given that it requires different people to be pulled in. Thanks!

**Comment by daxian-dbw** [comment]
@karelz You are right. I filed a different issue for the problem with the `Microsoft.CodeAnalysis.CSharp` package: #17797.
My question about the original issue has been answered/clarified. Thanks @karelz and @weshaggard!

**Comment by karelz** [comment]
Great, thanks!

@terrajobst we may still need to track some documentation/education on the topic ...

**Comment by hazzik** [comment]
>This is by design. They only support .NET Core, so if you want to consume them you need to do it from a netcoreapp library/application.

Then it shall be not possible to install.

**Comment by hazzik** [comment]
@weshaggard I disagree that this is "by design": all 3 packages target different versions of .NET Standard (1.5, 1.1 and 1.0 respectively), _not .NET Core App_. So they indeed should be compatible with .NET Standard 2.0. I believe that it shall be fixed in the same manner as #16801.

**Comment by weshaggard** [comment]
While their refs were targeting a version of netstandard, they were really only ever supported at runtime for .NET Core. So in this release we are no longer planning to ship an individual library package for them and they will only be available for .NET Core via Microsoft.NETCore.App package. 

We cannot do the same type of fix as #16801 because the implementations depend on the runtime which make them by nature not able to really build on top of the standard.
